### ISSUE TITLE
Create Slangpy mipmap learning sample

### ISSUE THREAD

**Issue by jarcherNV** [shader-slang/slang#6775]
Task is to create a mipmap learning example with Slangpy, for a Siggraph course.

General task breakdown:
- Create a slangpy application that renders a basic object with a texture + normal mapped texture. Mipmaps on.
- Have a way to swap the mipmap building algorithm for the normal map — none, simple averaging, and a  pre-saved version built with a classic normal mipmap builder eg toksvig approach
- Then, write a building algorithm for the normal map that is building based — Benedict can get description for you of the building algorithm to implement — but will be a slang side loss function, and then a rendering loop and probably though I’m not sure, Adam as the solver
- We will make a verison of this that is slang autodiff “free” but for now focus on buildign this with autodiff
- Chris can give example – Chris and Benedikt can give some help

**Comment by jarcherNV** [comment]
Some more info on what's needed for this after discussing with Benedikt:

When you generate mip maps, there are usually two options:
- Take the base texture, and downsample it with some basic image filtering (box/cubic/whichever). This works well for e.g. the albedo, but if you have a glossy material with a normal map, the material will look very different when rendered with the mips than with the base level. That's because image filtering knows nothing about reflection, and just downsampling normal/roughness textures will give bad results.
- The fix to this was to use a BRDF-aware downsampling method. The classic method for this is the Toksvig method from 20 years ago: https://developer.download.nvidia.com/whitepapers/2006/Mipmapping_Normal_Maps.pdf
- I don't expect you to reimplement this, but there are external normal mip map builders (nvidia definitely has one) that can produce these improved mipmaps
- These prior methods are nice, but they involves complicated math and don't work for all materials. The idea for this slangpy example is to show that you can do this with optimization. The outline is something like this:
- Take the base maps and produce mip 1 with some naive method (e.g. just box filter)
- Draw random samples (i.e. random uv, input direction, output direction) and evaluate the BRDF using mip 0 as input maps and using the (naive) mip 1.
- The values of the BRDF on the two different maps will be different (possibly very different), but we want them to looks as close as possible
- Take the difference between the two (e.g. using an L2 loss) and take the backwards derivative, propagating back to the (naive) mip 1. This will tell us how mip 1 needs to change to reduce the difference between renderings with mip 0 and mip 1
- Take a time step with the backpropagated gradients (e.g. using adam or just stochastic gradient descent)
- Repeat until convergence
- Now you have mipmapped normals/roughness/albedo/etc. for mip 1 that will render as close as possible to rendering with mip 0. From there, you can repeat the process to optimize mip 2 to look close to mip 1, etc.

**Comment by jarcherNV** [comment]
Update to this. I currently have a simple demo that generates a material and normal map texture using perlin noise which I'm using to calculate BRDF, and downsampled using a box filter.

I'm generating the textures using perlin noise for simplicity but I think next up I'd like to read the textures from a file instead. Then I still need to do the Toksvig stuff mentioned above, and then implement the learning part based on the BRDF difference.

**Comment by jarcherNV** [comment]
I have finished the initial commit for this and uploaded it to gitlab. Will need to get confirmation from Benedikt and Chris if this is ok or if it needs further changes, and then commit the final version perhaps to the Slangpy experiments.

**Comment by jarcherNV** [comment]
I have updated the WIP example in the dev/jarcher/mipmap-sample branch of the slangpy repo based on discussion from last week. Currently waiting for feedback from Benedikt regarding some issues I noticed. Specifically:
- When using a static light direction, it very quickly trains the normals to give a result that has almost 0 loss. With a changing light direction it trains more slowly, and while the loss is lower than (4) it never gets to 0 loss.
- This works when training from the downsampled normal map as inputs, but I also tried training a map of uniform normals (ie, all of them were 0,0,1) and this did not work. I'm not sure if that would be expected or if I've made a mistake in the training somewhere.
- Right now I'm just rendering the loss in (5) to confirm it's working. Alternatively I can render the BRDF result but it's kind of hard to see the difference. I'm wondering if this is expected.

**Comment by jarcherNV** [comment]
We have updated the sample to give a better visualisation of the training, and it now trains the normal map using an Adam optimizer rather than gradient descent, using random light and view directions starting with a set of uniform normals (all {0,0,1}) which gradually train over time to produce a more correct normal map.

Benedikt has suggested also training a roughness texture, for this we'll need to find a decent set of diffuse/normal/roughness textures that we can use.

**Comment by jarcherNV** [comment]
I have updated the sample to also train a roughness texture.

**Comment by jarcherNV** [comment]
While the sample is finished, it is currently sitting in an experimental dev branch. There should not be any more work for this task other than deciding if we want to merge the branch or not, so I am closing this as completed.
